Improve Python reference documentation

JeanChristopheMorinPerso opened the pull request:
**Link the Issue(s) this Pull Request is related to.**

Relates to #704 (and maybe others, I haven't searched extensively).

**Summarize your change.**

This PR refreshes the Python API reference documentation.

The rendered version can be found at https://jcmorin-opentimelineio.readthedocs.io/en/latest/index.html.

Changes are:
1. Addition of new docstrings (some from the previous Python implementation and some from me).
2. Move from running the `sphinx-apidoc` command to using `sphinx.ext.autosummary`. This gives better control over the generated documentation and improves the user experience.
3. The left menu in the docs is now organized in labelled categories. This is a small improvement, but it makes a huge readability/discover-ability improvements.
![Screenshot from 2022-02-06 11-30-26](https://user-images.githubusercontent.com/38703886/152690934-33c7d382-8310-4f06-b5f1-6ca28cdd5062.png)
4. Cleanup of the Sphinx configuration file to follow best practices established by the Sphinx community (only set the necessary settings, link to Sphinx docs instead of duplicating documentation of each setting, etc).
5. The docstrings generated by Pybind11 are now cleaned up by post-processors in Sphinx.
    * The `self` argument is stripped off as it's useless.
    * Overloads are properly handled and displayed in the rendered HTML.
6. Now using intersphinx to link to the python.org docs.
7. Switch the the newest ReadTheDocs config format.
8. Updated Python version used in ReadTheDocs to 3.10.
9. Added the spatial coordinates doc to the main doctree.
10. Simplified the build process a little bit.
11. Switched to https://myst-parser.readthedocs.io/en/latest/index.html since `recommonmark` is deprecated.
12. Dependencies needed to build the docs are now specified in `docs/requirements.txt`.
13. New GitHub Actions workflow that checks that all links are working and also checks that there is no warnings or errors when building the docs.

JeanChristopheMorinPerso commented on src/py-opentimelineio/opentimelineio/media_linker.py:
```diff
@@ -22,33 +22,37 @@
 
 Please raise an instance (or child instance) of
-otio.exceptions.CannotLinkMediaError() if there is a problem linking the media.
+:class:`.CannotLinkMediaError` if there is a problem linking the media.
```
`CannotLinkMediaError` is not defined in this repo... What should we do?

jminor replied:
We can remove this sentence.

JeanChristopheMorinPerso commented on src/py-opentimelineio/opentimelineio/core/__init__.py:
```diff
@@ -38,6 +38,32 @@
 )
 
+__all__ = [
+    'CannotComputeAvailableRangeError',
```
I had to make a choice here. This exception is made available in both `opentimelineio.exceptions` and `opentimelineio.core`.

I decided to publicly export it here to remind me about this. So here I am, exposing the problem. I'd prefer to have the exception exported in the `exceptions` module and remove the import from here, but that could break users...

jminor replied:
I think we intended all of the exceptions to be in the `opentimelineio.exceptions` module.

JeanChristopheMorinPerso commented on src/py-opentimelineio/opentimelineio/core/__init__.py:
```diff
@@ -38,6 +38,32 @@
     item,
 )
 
+__all__ = [
```
`__all__` is necessary to document the API as it should be. OTIO uses `from ... import ...` extensively to model the API. So to be representative of that in the reference documentation, we need `__all__`, else Sphinx wouldn't be able to document the class/function imports. As an example, see https://jcmorin-opentimelineio.readthedocs.io/en/latest/api/python/opentimelineio.core.html where all "core" classes are documented.

JeanChristopheMorinPerso commented on src/py-opentimelineio/opentimelineio/core/_core_utils.py:
```diff
@@ -228,6 +228,11 @@ def __deepcopy__(self, memo):
+                        continue
+
+                # Hide the method frm Sphinx doc. See https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#info-field-lists
+                getattr(mapClass, name).__doc__ += '\n\n:meta private:'
```
Little trick to unclutter the reference documentation with unnecessary (well at least IMO) methods.

JeanChristopheMorinPerso commented on src/py-opentimelineio/opentimelineio/core/composition.py:
```diff
@@ -39,19 +39,19 @@ def each_child(
 
-    Note that this function is now deprecated, please consider using
-    children_if() instead.
+    .. deprecated:: 0.14.0
```
These are now properly marked as deprecated in the docs.

jhodges10 commented on src/py-opentimelineio/opentimelineio/schema/timeline.py:
```diff
@@ -19,33 +19,31 @@ def __repr__(self):
 @add_method(_otio.Timeline)
 def each_child(self, search_range=None, descended_from_type=_otio.Composable):
-    """ Generator that returns each child contained in the timeline
+    """ enerator that returns each child contained in the timeline
```
I don't think you meant to delete the `G` here, did you?

JeanChristopheMorinPerso replied:
Nice catch. I wanted to delete the leading space, not the "G".

jminor commented on src/py-opentimelineio/opentime-bindings/opentime_rationalTime.cpp:
```diff
@@ -57,19 +57,25 @@ RationalTime _type_checked(py::object const& rhs, char const* op) {
+Returns true if the time is invalid. The time is considered invalid if the value or the rate are a NaN value
+or if the value of the rate is smaller or equal to zero.
+)docstring")
+        .def_property_readonly("value", &RationalTime::value, "This is the value property doc")
```
```suggestion
        .def_property_readonly("value", &RationalTime::value)
```

jminor commented on src/py-opentimelineio/opentime-bindings/opentime_rationalTime.cpp:
```diff
@@ -78,25 +84,33 @@ void opentime_rationalTime_bindings(py::module m) {
+
+For example, the duration of a clip from frame 10 to frame 15 is 6 frames. Result will be in the rate of start_time.
+)docstring")
+        .def_static("is_valid_timecode_rate", &RationalTime::is_valid_timecode_rate, "rate"_a, "Returns true is the rate is a known valid timecode.")
```
```suggestion
        .def_static("is_valid_timecode_rate", &RationalTime::is_valid_timecode_rate, "rate"_a, "Returns true if the rate is valid for use with timecode.")
```

jminor commented on src/py-opentimelineio/opentime-bindings/opentime_rationalTime.cpp:
```diff
@@ -78,25 +84,33 @@ void opentime_rationalTime_bindings(py::module m) {
-            R"docstring(Returns the first valid timecode rate that has the least difference from the given value.)docstring")
-        .def_static("from_frames", &RationalTime::from_frames, "frame"_a, "rate"_a)
+            "Returns the first valid timecode rate that has the least difference from the given value.")
+        .def_static("from_frames", &RationalTime::from_frames, "frame"_a, "rate"_a, "Turn a frame number and rate into a time object.")
```
```suggestion
        .def_static("from_frames", &RationalTime::from_frames, "frame"_a, "rate"_a, "Turn a frame number and rate into a :class:`~RationalTime` object.")
```

jminor commented on src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp:
```diff
@@ -199,7 +199,7 @@ static void define_bases2(py::module m) {
 
     py::class_<Composable, SOWithMetadata,
-               managing_ptr<Composable>>(m, "Composable", py::dynamic_attr())
+               managing_ptr<Composable>>(m, "Composable", py::dynamic_attr(), "An object that can be composed by :class:`~Track`\\s.")
```
```suggestion
               managing_ptr<Composable>>(m, "Composable", py::dynamic_attr(), "An object that can be composed within a :class:`~Composition` (such as :class:`~Track` or :class:`~Stack`).")
```

jminor commented on src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp:
```diff
@@ -211,7 +211,7 @@ static void define_bases2(py::module m) {
 
     auto marker_class =
-        py::class_<Marker, SOWithMetadata, managing_ptr<Marker>>(m, "Marker", py::dynamic_attr())
+        py::class_<Marker, SOWithMetadata, managing_ptr<Marker>>(m, "Marker", py::dynamic_attr(), "Holds metadata over time on a timeline")
```
```suggestion
        py::class_<Marker, SOWithMetadata, managing_ptr<Marker>>(m, "Marker", py::dynamic_attr(), "A marker indicates a moment or range of time on an item in a timeline, usually with a name, color or other metadata.")
```

jminor commented on src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp:
```diff
@@ -250,7 +250,13 @@ static void define_bases2(py::module m) {
     py::class_<SerializableCollection, SOWithMetadata,
-               managing_ptr<SerializableCollection>>(m, "SerializableCollection", py::dynamic_attr())
+               managing_ptr<SerializableCollection>>(m, "SerializableCollection", py::dynamic_attr(), R"docstring(
+A kind of composition which can hold any serializable object.
```
```suggestion
A container which can hold an ordered list of any serializable objects. Note that this is not a :class:`~Composition` nor is it :class:`~Composable`.
```

jminor commented on src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp:
```diff
@@ -250,7 +250,13 @@ static void define_bases2(py::module m) {
+               managing_ptr<SerializableCollection>>(m, "SerializableCollection", py::dynamic_attr(), R"docstring(
+A kind of composition which can hold any serializable object.
+
+This composition approximates the concept of a bin - a collection of :class:`.SerializableObject`\s that do
```
```suggestion
This container approximates the concept of a bin - a collection of :class:`.SerializableObject`\s that do
```

jminor commented on src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp:
```diff
@@ -250,7 +250,13 @@ static void define_bases2(py::module m) {
+A kind of composition which can hold any serializable object.
+
+This composition approximates the concept of a bin - a collection of :class:`.SerializableObject`\s that do
+not have any compositing meaning, but can serialize to/from OTIO correctly, with metadata and
```
```suggestion
not have any compositional meaning, but can serialize to/from OTIO correctly, with metadata and
```

jminor commented on src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp:
```diff
@@ -250,7 +250,13 @@ static void define_bases2(py::module m) {
+
+This composition approximates the concept of a bin - a collection of :class:`.SerializableObject`\s that do
+not have any compositing meaning, but can serialize to/from OTIO correctly, with metadata and
+a named collection.
```
A :class:`~SerializableCollection` is useful for serializing multiple timelines, clips, or media references to a single file.

jminor commented on src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp:
```diff
@@ -307,7 +313,7 @@ static void define_items_and_compositions(py::module m) {
              "enabled"_a = true,
              py::arg_v("metadata"_a = py::none()))
-        .def_property("enabled", &Item::enabled, &Item::set_enabled, "If true, an Item contributes to compositions. Analogous to Mute in various NLEs.")
+        .def_property("enabled", &Item::enabled, &Item::set_enabled, "If true, an Item contributes to compositions. Analogous to 'mute' in various NLEs.")
```
```suggestion
        .def_property("enabled", &Item::enabled, &Item::set_enabled, "If true, an Item contributes to compositions. For example, when an audio/video clip is `enabled=false` the clip is muted/hidden.")
```

jminor commented on src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp:
```diff
@@ -341,7 +347,7 @@ static void define_items_and_compositions(py::module m) {
 
     auto transition_class =
-        py::class_<Transition, Composable, managing_ptr<Transition>>(m, "Transition", py::dynamic_attr())
+        py::class_<Transition, Composable, managing_ptr<Transition>>(m, "Transition", py::dynamic_attr(), "Represents a transition between two items.")
```
```suggestion
        py::class_<Transition, Composable, managing_ptr<Transition>>(m, "Transition", py::dynamic_attr(), "Represents a transition between the two adjacent items in a :class:`~Track`. For example, a cross dissolve or wipe.")
```

jminor commented on src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp:
```diff
@@ -396,7 +408,11 @@ static void define_items_and_compositions(py::module m) {
+    py::class_<Clip, Item, managing_ptr<Clip>>(m, "Clip", py::dynamic_attr(), R"docstring(
+The base editable object in OTIO.
+
+Contains a :class:`.MediaReference` and a trim on that media reference.
```
Is the "." before MediaReference meant to be a "~"? I'm not familiar with this syntax but I see most of the others are "~"

JeanChristopheMorinPerso replied:
If you are curious, here is two links explaining cross-references in Sphinx:
* https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#cross-referencing-syntax
* https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#cross-referencing-python-objects

Sometimes it's necessary to use `.Something` and sometimes `~Something` is enough.

jminor commented on src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp:
```diff
@@ -396,7 +408,11 @@ static void define_items_and_compositions(py::module m) {
 
-    py::class_<Clip, Item, managing_ptr<Clip>>(m, "Clip", py::dynamic_attr())
+    py::class_<Clip, Item, managing_ptr<Clip>>(m, "Clip", py::dynamic_attr(), R"docstring(
+The base editable object in OTIO.
```
```suggestion
A :class:`~Clip` is a segment of editable media (usually audio or video).
```

jminor commented on src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp:
```diff
@@ -611,7 +631,7 @@ static void define_effects(py::module m) {
         .def_property("effect_name", &Effect::effect_name, &Effect::set_effect_name);
 
-    py::class_<TimeEffect, Effect, managing_ptr<TimeEffect>>(m, "TimeEffect", py::dynamic_attr())
+    py::class_<TimeEffect, Effect, managing_ptr<TimeEffect>>(m, "TimeEffect", py::dynamic_attr(), "Base Time Effect Class")
```
```suggestion
    py::class_<TimeEffect, Effect, managing_ptr<TimeEffect>>(m, "TimeEffect", py::dynamic_attr(), "Base class for all effects that alter the timing of an item.")
```

jminor commented on src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp:
```diff
@@ -620,7 +640,9 @@ static void define_effects(py::module m) {
 
-    py::class_<LinearTimeWarp, TimeEffect, managing_ptr<LinearTimeWarp>>(m, "LinearTimeWarp", py::dynamic_attr())
+    py::class_<LinearTimeWarp, TimeEffect, managing_ptr<LinearTimeWarp>>(m, "LinearTimeWarp", py::dynamic_attr(), R"docstring(
+A time warp that applies a linear scale across the entire clip
```
```suggestion
A time warp that applies a linear speed up or slow down across the entire clip.
```

jminor commented on src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp:
```diff
@@ -629,9 +651,9 @@ static void define_effects(py::module m) {
              "time_scalar"_a = 1.0,
              py::arg_v("metadata"_a = py::none()))
-        .def_property("time_scalar", &LinearTimeWarp::time_scalar, &LinearTimeWarp::set_time_scalar);
+        .def_property("time_scalar", &LinearTimeWarp::time_scalar, &LinearTimeWarp::set_time_scalar, "Linear time scalar applied to clip. 2.0 = double speed, 0.5 = half speed.");
```
Note that adjusting the time_scalar of a LinearTimeWarp does not affect the duration of the item this effect is attached to. Instead it affects the speed of the media displayed within that item.

jminor commented on src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp:
```diff
@@ -672,7 +694,7 @@ static void define_media_references(py::module m) {
 
     py::class_<MissingReference, MediaReference,
-               managing_ptr<MissingReference>>(m, "MissingReference", py::dynamic_attr())
+               managing_ptr<MissingReference>>(m, "MissingReference", py::dynamic_attr(), "Represents media for which a concrete reference is missing.")
```
Note that a MissingReference may have useful metadata, even if the location of the media is not known.

jminor commented on src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp:
```diff
@@ -806,12 +834,12 @@ Negative ``start_frame`` is also handled. The above example with a ``start_frame
-        .def_property("missing_frame_policy", &ImageSequenceReference::missing_frame_policy, &ImageSequenceReference::set_missing_frame_policy, "Enum ``ImageSequenceReference.MissingFramePolicy`` directive for how frames in sequence not found on disk should be handled.")
-        .def("end_frame", &ImageSequenceReference::end_frame, "Last frame number in the sequence based on the ``rate`` and ``available_range``.")
-        .def("number_of_images_in_sequence", &ImageSequenceReference::number_of_images_in_sequence, "Returns the number of images based on the ``rate`` and ``available_range``.")
+        .def_property("missing_frame_policy", &ImageSequenceReference::missing_frame_policy, &ImageSequenceReference::set_missing_frame_policy, "Directive for how frames in sequence not found on disk should be handled.")
```
```suggestion
        .def_property("missing_frame_policy", &ImageSequenceReference::missing_frame_policy, &ImageSequenceReference::set_missing_frame_policy, "Directive for how frames in sequence not found during playback or rendering should be handled.")
```

jminor reviewed (changes requested):
This is amazing! I made some edits in several places that will hopefully make some things more clear.

JeanChristopheMorinPerso commented:
Thanks @jminor for the review! I addressed all your comments and suggestions and I also completed what I wanted to do. The ReadTheDocs setup should be much more reliable now because the dependencies required to build the docs are now pinned to specific versions.

JeanChristopheMorinPerso commented:
It's also good to note that their is only 2 Sphinx warnings left when building the docs :)

```
docstring of opentimelineio._otio.PyCapsule.parent:: WARNING: py:class reference target not found: opentimelineio::v1_0::Composition
docstring of opentimelineio._otio.Item:: WARNING: py:class reference target not found: bool_
```
To remove the first warning, we would need to shuffle the order of the code in `src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp`.

For the second warning, we would need to update to Pybind11 [2.9.1](https://github.com/pybind/pybind11/releases/tag/v2.9.1) where we are interested in `Render py::bool_ and py::float_ as bool and float respectively.`.

JeanChristopheMorinPerso commented on src/py-opentimelineio/opentimelineio/core/__init__.py:
```diff
@@ -2,9 +2,6 @@
 
 from .. _otio import ( # noqa
-    # errors
-    CannotComputeAvailableRangeError,
```
This is a potentially breaking change.

meshula replied:
you mean breaking for third parties, if they use it? Maybe open an issue as an RFC linked to this PR and if no one speaks up we can (1) document that it's removed in the change notes, (2) decide that the removal is fine.

JeanChristopheMorinPerso replied:
I added it back but it won't show up in the API reference. I will open a separate issue to remove it.

JeanChristopheMorinPerso commented:
Alright, I fixed the `docstring of opentimelineio._otio.PyCapsule.parent:: WARNING: py:class reference target not found: opentimelineio::v1_0::Composition` warning. Only one left, but I'll do that in another pull request since it implies updating Pybind11.

JeanChristopheMorinPerso commented on src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp:
```diff
@@ -203,20 +203,8 @@ static void define_bases2(py::module m) {
     MarkerVectorProxy::define_py_class(m, "MarkerVector");
     EffectVectorProxy::define_py_class(m, "EffectVector");
 
-    py::class_<Composable, SOWithMetadata,
```
Moved into the `define_items_and_compositions` function since the `parent` method returns a Composition which is defined in `define_items_and_compositions`.

meshula commented on docs/conf.py:
```diff
@@ -2,306 +2,172 @@
+
+# -- Custom ----------------------------------------------------------------------------
+
+def process_signature(
```
does the need for process_signature disappear completely when 2621 lands? or just the identified chunk below?

JeanChristopheMorinPerso replied:
Just the identified block (https://github.com/PixarAnimationStudios/OpenTimelineIO/pull/1209/files#diff-85933aa74a2d66c3e4dcdf7a9ad8397f5a7971080d34ef1108296a7c6b69e7e3R125-R144). Removing `self` is still something that is not yet talked about in Pybind11 issues.

meshula commented on docs/conf.py:
```diff
@@ -2,306 +2,172 @@
+    return signature, return_annotation
+
+
+def process_docstring(
```
same question ~ I assume this routine as well is just about removing .self.

meshula replied:
or do these routines serve to remove the need for api_doc?

JeanChristopheMorinPerso replied:
This one removes self from the docstrings in case where a function/method is overloaded. For example, see https://jcmorin-opentimelineio.readthedocs.io/en/latest/api/python/opentimelineio.core.html#opentimelineio.core.flatten_stack. The docstring doesn't contain `self`  because it was removed by `process_docstring`.

So I guess I need to add a little explanation for this one.

JeanChristopheMorinPerso replied:
I added a comment in the function to clarify.

meshula commented on docs/tutorials/otio-serialized-schema.md:
```diff
@@ -455,14 +482,19 @@ parameters:
 - *metadata*: 
 - *name*: 
-- *time_scalar*: 
+- *time_scalar*: Linear time scalar applied to clip. 2.0 = double speed, 0.5 = half speed.
```
```suggestion
- *time_scalar*: Linear time scalar applied to clip. 2.0 means the clip occupies half the time in the parent item, i.e. plays at double speed, 0.5 means the clip occupies twice the time in the parent item, i.e. plays at half speed.
```

meshula commented on docs/tutorials/otio-serialized-schema.md:
````diff
@@ -471,12 +503,12 @@ parameters:
 
 ```
-None
+A marker indicates a moment or range of time on an item in a timeline, usually with a name, color or other metadata.
````
```suggestion
A marker indicates a marked range of time on an item in a timeline, usually with a name, color or other metadata. The marked range may have a zero duration. The marked range is in the owning item's time coordinate system.
```

meshula commented on docs/tutorials/quickstart.md:
```diff
@@ -9,7 +9,7 @@ This is for users who wish to get started using the "OTIOView" application to in
 
 - Try `python -m pip install PySide2` or `python -m pip install PySide6`
-- If that doesn't work, try downloading PySide here: <a href="https://wiki.qt.io/Qt_for_Python" target="_blank">https://wiki.qt.io/Qt_for_Python</a>
+- If that doesn't work, try downloading PySide here: [](https://wiki.qt.io/Qt_for_Python)
```
```suggestion
- If difficulties are encountered, please file an issue on OpenTimelineIO's github for assistance.
```

meshula commented on docs/tutorials/quickstart.md:
```diff
@@ -9,7 +9,7 @@ This is for users who wish to get started using the "OTIOView" application to in
-- If that doesn't work, try downloading PySide here: <a href="https://wiki.qt.io/Qt_for_Python" target="_blank">https://wiki.qt.io/Qt_for_Python</a>
+- If that doesn't work, try downloading PySide here: [](https://wiki.qt.io/Qt_for_Python)
 
 You probably want the prebuilt binary for your platform.  PySide generally includes a link to the appropriate version of Qt as well.
```
```suggestion
```

meshula commented on docs/tutorials/quickstart.md:
```diff
@@ -20,7 +20,7 @@ You probably want the prebuilt binary for your platform.  PySide generally inclu
 ## Configure Environment Variables for extra adapters
 
 By default, when you install OTIO you will only get the "Core" adapters, which include CMX EDL, Final Cut Pro 7 XML, and the built in JSON format.  In order to get access to the "contrib" adapters (which includes the maya sequencer, rv and others) you'll need to set some environment variables.  If you need support for these formats, please consult the 
```
```suggestion
A default OTIO installation includes only the "Core" adapters, which include CMX EDL, Final Cut Pro 7 XML, and the built in JSON format.  In order to get access to the "contrib" adapters (which includes the maya sequencer, rv and others), please consult the 
```

meshula commented on src/py-opentimelineio/opentime-bindings/opentime_rationalTime.cpp:
```diff
@@ -116,10 +130,10 @@ void opentime_rationalTime_bindings(py::module m) {
         .def_static("from_time_string", [](std::string s, double rate) {
                 return RationalTime::from_time_string(s, rate, ErrorStatusConverter());
-            }, "time_string"_a, "rate"_a)
+            }, "time_string"_a, "rate"_a, "Convert a time with microseconds string (``HH:MM:ss.ms``) into a :class:`~RationalTime`.")
```
```suggestion
            }, "time_string"_a, "rate"_a, "Convert a time with microseconds string (``HH:MM:ss`` where ss is an integer or a decimal number) into a :class:`~RationalTime`.")
```

meshula commented on src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp:
```diff
@@ -362,21 +363,26 @@ static void define_items_and_compositions(py::module m) {
+    py::class_<Transition::Type>(transition_class, "Type", R"docstring(
+Enum encoding types of transitions.
 
+This is for representing “Dissolves” and “Wipes” defined by the multi-source effect as defined in section 7.6.3.2 of `SMPTE 258M-2004 <https://ieeexplore.ieee.org/servlet/opac?punumber=7291837>`_.
```
That document has restricted, and paywalled access. Do we want to point to it? I argue no. Also the enumeration in OTIO has only dissolve. I don't think we want to formalize the SMPTE set of wipes without discussion. The majority of them are exotic analog effects from the mid to late twentieth century, such as "wiggly barndoor zigzag", "spiralling inverted bananas (counter clockwise)", and "I'm an intern at a public broadcasting station and no one is looking hahahaha". If we do adopt an enumeration of more than one element, we should discuss at the TSC.

JeanChristopheMorinPerso replied:
I removed the sentence.

meshula commented on src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp:
```diff
@@ -651,9 +678,14 @@ static void define_effects(py::module m) {
              py::arg_v("metadata"_a = py::none()))
-        .def_property("time_scalar", &LinearTimeWarp::time_scalar, &LinearTimeWarp::set_time_scalar);
+        .def_property("time_scalar", &LinearTimeWarp::time_scalar, &LinearTimeWarp::set_time_scalar, R"docstring(
+Linear time scalar applied to clip. 2.0 = double speed, 0.5 = half speed.
```
see above for suggested edit

meshula commented on src/py-opentimelineio/opentimelineio/adapters/fcp_xml.py:
```diff
@@ -561,9 +561,7 @@ class FCP7XMLParser:
-    .. seealso:: https://developer.apple.com/library/archive/documentation/\
-            AppleApplications/Reference/FinalCutPro_XML/Basics/Basics.html\
-            #//apple_ref/doc/uid/TP30001154-TPXREF102
+    .. seealso:: https://developer.apple.com/library/archive/documentation/AppleApplications/Reference/FinalCutPro_XML/Basics/Basics.html#//apple_ref/doc/uid/TP30001154-TPXREF102 # noqa
```
is seealso a typo or a keyword?

JeanChristopheMorinPerso replied:
Unfortunately it's a keyboard with a typo in it, see https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-seealso.

meshula commented on src/py-opentimelineio/opentimelineio/hooks.py:
```diff
@@ -9,58 +9,71 @@
 manifest.
 
-You can attach multiple hook scripts to a hook.  They will be executed in list
+You can attach multiple hook scripts to a hook. They will be executed in list
```
```suggestion
Multiple scripts can be attached to a hook. They will be executed in list
```

meshula commented on src/py-opentimelineio/opentimelineio/opentime.py:
```diff
@@ -29,12 +48,17 @@ def to_timecode(rt, rate=None, drop_frame=None):
 
 
 def to_time_string(rt):
+    """Convert this timecode to time with microsecond, as formated in
```
note that the documentation on that page is incorrect and misleading. We should not link it.

The correct information is here (thanks to @jchen9 for the reference) ~ https://github.com/FFmpeg/FFmpeg/blob/83e1a1de8833845224948e5d002355c03dd117d5/fftools/ffprobe.c#L412

```suggestion
    """Convert this timecode to time as used by ffmpeg, formatted as hh:mm:ss where ss is an integer or decimal number
```

JeanChristopheMorinPerso replied:
Would https://ffmpeg.org/ffprobe.html#Main-options be a better documentation?

```
See the [documentation of the -sexagesimal flag of ffprobe](https://ffmpeg.org/ffprobe.html#Main-options)
```
?
Or would you prefer that we simply don't link to anything and just describe the format OTIO accepts?

meshula replied:
The documentation at ffmpeg.org is incorrect there as well. It says seconds.milliseconds, but the code in ffprobe simply says %f with seconds as the argument. Their documentation explicitly says 3.5 means 3 and a half seconds, not 3 seconds and 5 milliseconds. I'm comfortable with stating this value, as interpretedby otio is compatible with ffprobe, but not comfortable at pointing to a misleading explanation. That's why I rewrote the explanations here as "ss means an integer or a decimal number" because that's what the ffprobe code does. 

As always, if I've misread the code, I'd appreciate correction, because I don't want to insist on something incorrect. I'm confused by the apparent mismatch of the documentation and the code, frankly. In general I haven't seen this kind of problem with ffmpeg before.

JeanChristopheMorinPerso replied:
So I simply went with your suggestion.

meshula commented on src/py-opentimelineio/opentimelineio/opentime.py:
```diff
@@ -29,12 +48,17 @@ def to_timecode(rt, rate=None, drop_frame=None):
 
 def to_time_string(rt):
+    """Convert this timecode to time with microsecond, as formated in
+    `FFMPEG <https://trac.ffmpeg.org/wiki/Seeking#Timeunitsyntax>`_
```
```suggestion
    Please note that the documentation here does not correspond to the implementation of ffmpeg: `FFMPEG <https://trac.ffmpeg.org/wiki/Seeking#Timeunitsyntax>`_
```

JeanChristopheMorinPerso replied:
I just removed the reference to this doc.

JeanChristopheMorinPerso commented:
I pushed new commits that address the PR comments I had from @meshula and there is also a new GitHub action workflow that checks for dead links in the docs.

I had to rebase all commits because of DCO and fix merge conflicts.

JeanChristopheMorinPerso commented on docs/tutorials/adapters.md:
```diff
@@ -3,48 +3,47 @@
 - Full specification: SMPTE 258M-2004 "For Television −− Transfer of Edit Decision Lists"
 - http://xmil.biz/EDL-X/CMX3600.pdf
-- <a href="https://documentation.apple.com/en/finalcutpro/usermanual/index.html#chapter=96%26section=1" target="_blank">Reference</a>
+- [Reference](https://prohelp.apple.com/finalcutpro_help-r01/English/en/finalcutpro/usermanual/chapter_96_section_0.html)
```
This link is now fixed. It was dead before and @jminor suggested an internet archive link. But after some searching, I found this link that is "exactly" what was on the internet archive. I used the section 0 instead of 1, but that's just because of the documentation layout that changed slightly. The content is still the same as far as I see.

JeanChristopheMorinPerso commented:
I create a PR to update Pybind11 to 2.9.2, see #1312 . Hopefully it will be merged before we merge this current PR so that I can add a new job to check for Sphinx warnings in CI.

JeanChristopheMorinPerso commented:
@jminor @meshula (I wish there was a TSC tag... There is only the committer tag which includes John Mertic...), Gentle ping on this PR. I just added the job that checks if there is any build error/warning as I said previously.

This PR is now pretty much done from my point of view, unless someone wants to do another review pass at the docstrings or anything else?

JeanChristopheMorinPerso commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,105 @@
+        if [ $exit_code -eq 0 ]; then
+            echo -e "\n\n=================\nAll links are valid!"
+
+            echo "# :heavy_check_mark: Sphinx links" >> $GITHUB_STEP_SUMMARY
```
You can see the result of this in https://github.com/JeanChristopheMorinPerso/OpenTimelineIO/actions/runs/2521426263/attempts/1#summary-6950080420 when it succeeds and https://github.com/JeanChristopheMorinPerso/OpenTimelineIO/actions/runs/2521417320/attempts/1#summary-6950061883 when it fails. I thought it would be a nice addition that can be later improved even more.

ssteinbach commented:
This looks great, thanks @JeanChristopheMorinPerso !  One (non blocking) question- do you know why some headers appear twice on mobile?

![image](https://user-images.githubusercontent.com/61017/174491801-46496986-e157-4309-b125-89d7bf9e6603.jpeg)

JeanChristopheMorinPerso commented:
Good question @ssteinbach ! Both pages seem to contain the same section names. When you click on the sections, the content is different, so they don't seem duplicated.

Does that make sense?

ssteinbach reviewed (approved):
Wow this is a great change. Thanks for working through this!

JeanChristopheMorinPerso commented:
Thanks for the reviews and accepting the pull request! 🎉✨
